protocols/kad/src/behaviour.rs

```diff
@@ -1475,9 +1509,26 @@ where
                         key, records, quorum, cache_at
                     } = &mut query.inner.info {
                         if let Some(record) = record {
-                            records.push(record);
+                            records.push(PeerRecord{ peer: Some(source.clone()), record });
+
                             if records.len() >= quorum.get() {
-                                query.finish()
+                                // Desired quorum reached. The query may finish. See
+                                // [`Query::try_finish`] for details.
+                                let peers = records.iter()
+                                    .filter_map(|PeerRecord{ peer, .. }| peer.as_ref())
+                                    .cloned()
+                                    .collect::<Vec<_>>();
+                                let is_finished = query.try_finish(peers.iter());
+
+                                debug!(
+                                    "GetRecord query with id {:?} reached the quorum {} finished \
+                                     with recent response from peer {}, counting {} successful \
+                                     responses total.",
+                                    user_data,
+                                    if is_finished { "and is" } else { "but is not yet" },
+                                    source,
+                                    peers.len(),
+                                );
```

Comment:
```suggestion
                                let finished = query.try_finish(peers.iter());
                                if !finished {
                                    debug!(
                                        "GetRecord query ({:?}) reached quorum ({}/{}) \
                                         with response from peer {} but could not yet finish. 
                                        user_data, peers.len(), quorum.get(),
                                        source,
                                    );
                                }
```
I think the only interesting case to log here is if the query is not yet finished. There is some general logging for finished/timed out queries already in `query_finished` and `query_timeout`.